gapic/samplegen/samplegen.py

```diff
@@ -434,52 +449,82 @@ def validate_expression(self, exp: str) -> wrappers.Field:
         Returns:
             wrappers.Field: The final field in the chain.
         """
-        # TODO: handle mapping attributes, i.e. {}
         # TODO: Add resource name handling, i.e. %
-        indexed_exp_re = re.compile(
-            r"^(?P<attr_name>\$?\w+)(?:\[(?P<index>\d+)\])?$")
-
-        toks = exp.split(".")
-        match = indexed_exp_re.match(toks[0])
-        if not match:
-            raise BadAttributeLookup(
-                f"Badly formatted attribute expression: {exp}")
-
-        base_tok, previous_was_indexed = (match.groupdict()["attr_name"],
-                                          bool(match.groupdict()["index"]))
-        base = self.var_field(base_tok)
-        if not base:
-            raise UndefinedVariableReference(
-                "Reference to undefined variable: {}".format(base_tok))
-        if previous_was_indexed and not base.repeated:
-            raise BadAttributeLookup(
-                "Cannot index non-repeated attribute: {}".format(base_tok))
-
-        for tok in toks[1:]:
-            match = indexed_exp_re.match(tok)
+        chain_link_re = re.compile(
+            r"""
+            (?P<attr_name>\$?\w+)(?:\[(?P<index>\d+)\]|\{["'](?P<key>[^"']+)["']\})?$
+            """.strip())
+
+        def validate_recursively(expression, scope, depth=0):
+            first_dot = expression.find(".")
+            base = expression[:first_dot] if first_dot > 0 else expression
+            match = chain_link_re.match(base)
             if not match:
                 raise BadAttributeLookup(
-                    f"Badly formatted attribute expression: {tok}")
+                    f"Badly formed attribute expression: {expression}")
+
+            name, idxed, mapped = (match.groupdict()["attr_name"],
+                                   bool(match.groupdict()["index"]),
+                                   bool(match.groupdict()["key"]))
```

Comment:
Yes, that's correct. The regex groups save the values of `index` or `key` purely for debugging, and because at some point in the future they may be useful.